napari/_vispy/layers/labels.py

```diff
@@ -478,24 +511,27 @@ def _on_colormap_change(self, event=None):
         mode = self.layer.color_mode
 
         if mode == 'auto':
-            dtype = minimum_dtype_for_labels(self.layer.num_colors + 1)
-            if issubclass(dtype.type, np.integer):
-                scale = np.iinfo(dtype).max
-            else:  # float32 texture
-                scale = 1.0
+            data_dtype = self.layer._slice.image.view.dtype
+            raw_dtype = self.layer._slice.image.raw.dtype
+            if data_dtype != raw_dtype:
+                colormap = LabelColormap(**colormap.dict())
```

Comment:
(I don't know if I understand this, but correct me if my comment is mistaken.)

```suggestion
                # If the texture dtype and the data dtype don't match,
                # we need to make a copy of the colormap with a
                # potentially different background value
                colormap = LabelColormap(**colormap.dict())
```

Also, I note that for the `LabelVispyColormap`, these variables are called `view_dtype` and `data_dtype`, and these correspond in this function to `data_dtype` and `raw_dtype`, respectively. This is confusing! We need to settle on a terminology. I suggest `texture_dtype` and `data_dtype`. Does that capture the issue well enough or is there another intermediary value?

Reply:
I have added own comment with explanation